presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java

```diff
@@ -40,23 +41,38 @@
     private final List<SortingColumn> sortedBy;
     private final BucketFunctionType bucketFunctionType;
     private final Optional<List<Type>> types;
+    private final Optional<List<String>> clusteredBy;
+    private final Optional<List<Integer>> clusterCount;
+    private final Optional<List<String>> distribution;
 
     @JsonCreator
     public HiveBucketProperty(
             @JsonProperty("bucketedBy") List<String> bucketedBy,
             @JsonProperty("bucketCount") int bucketCount,
             @JsonProperty("sortedBy") List<SortingColumn> sortedBy,
             @JsonProperty("bucketFunctionType") BucketFunctionType bucketFunctionType,
-            @JsonProperty("types") Optional<List<Type>> types)
+            @JsonProperty("types") Optional<List<Type>> types,
+            @JsonProperty("clusteredBy") Optional<List<String>> clusteredBy,
+            @JsonProperty("clusterCount") Optional<List<Integer>> clusterCount,
+            @JsonProperty("distribution") Optional<List<String>> distribution
+            )
     {
         this.bucketedBy = ImmutableList.copyOf(requireNonNull(bucketedBy, "bucketedBy is null"));
         this.bucketCount = bucketCount;
         this.sortedBy = ImmutableList.copyOf(requireNonNull(sortedBy, "sortedBy is null"));
         this.bucketFunctionType = requireNonNull(bucketFunctionType, "bucketFunctionType is null");
         this.types = requireNonNull(types, "type is null");
+        this.clusteredBy = requireNonNull(clusteredBy, "clusteredBy is null");
+        this.clusterCount = requireNonNull(clusterCount, "clusterCount is null");
+        this.distribution = requireNonNull(distribution, "distribution is null");
+
         if (bucketFunctionType.equals(PRESTO_NATIVE)) {
             checkArgument(types.isPresent(), "Types must be present for bucket function type " + bucketFunctionType);
             checkArgument(types.get().size() == bucketedBy.size(), "The sizes of bucketedBy and types should match");
+        } else if (bucketFunctionType.equals(HIVE_CLUSTERING)) {
+            checkArgument(clusteredBy.isPresent(), "Cluster columns must be present for bucket function type " + bucketFunctionType);
+            checkArgument(clusterCount.isPresent(), "Cluster numbers must be present for bucket function type " + bucketFunctionType);
+            checkArgument(bucketedBy.isEmpty(), "Bucket columns should not be present for bucket function type " + bucketFunctionType);
```

Comment:
I guess clustering is compatible with sortedBy but want to call out here in case I'm not understanding correctly :)

Reply:
@shixuan-fan, right! Clustering should be naturally compatible with SortedBy. I will add tests to confirm  if it works.